platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,998 @@
+            .onAbort(this::logInterruptionAndReconnectIfStillConnecting)
+            .onFailure(this::logAndReconnectIfNotCancelledAndStillConnecting)
+            .onSuccess(this::logAndSetConnected);
+    this.waiter = waiter;
+    this.executor = executor;
+    this.lastPing = new AtomicLong(System.currentTimeMillis());
+    this.lastCreateError =
+        new AtomicLong(); // set to beginning of time to make sure we log the first error
+    this.unavailableClient =
+        new UnavailableSolrClient(
+            new UnavailableSolrException("initializing '" + core + "' client"));
+    this.apiClient = unavailableClient;
+    this.pingClient = unavailableClient;
+    this.realClient = null;
+    this.state = State.CREATING;
+    this.future = null;
+    setCreating(unavailableClient, false);
+  }
+
+  @Override
+  protected SolrClient getProxiedClient() {
+    if ((state != State.CONNECTED) && (state != State.CLOSED)) {
+      // do a spot check to see if it suddenly became reachable
+      try {
+        checkIfReachable("from the API because it is currently unavailable");
+        // if we get here than the ping was successful so make sure we move to connected
+        setConnected(true);
+        // fall-through to return the current one which should have been changed to the actual one
+      } catch (UnavailableSolrException e) {
+        // fall-through to return the current one which should still be an unavailable one
+      }
+    }
+    return this.apiClient;
+  }
+
+  @Override
+  protected <T> T handle(Code<T> code) throws SolrServerException, IOException {
+    try {
+      return code.invoke(getProxiedClient());
+    } catch (UnavailableSolrException e) {
+      throw e;
+    } catch (SolrException e) {
+      LOGGER.debug(
+          "Solr({}): API failure with code [{}] and metadata [{}]; {}",
+          core,
+          e.code(),
+          e.getMetadata(),
+          e, // this will get the info logged on the first line
+          e); // this one will get the stack trace logged after
+      checkIfReachableAndChangeStateAccordingly(e, "from the API after an error was detected");
+      throw e;
+    } catch (SolrServerException | IOException e) {
+      LOGGER.debug("Solr({}): API failure; {}", core, e, e);
+      checkIfReachableAndChangeStateAccordingly(e, "from the API after an error was detected");
+      throw e;
+    }
+  }
+
+  @Override
+  public final SolrClient getClient() {
+    // returning this to make sure all calls from our SolrClient API will still be intercepted
+    // when the client is retrieved and passed to a Solr request object
+    // this should be temporary until we completely abstract out Solr with interfaces
+    return this;
+  }
+
+  @Override
+  public String getCore() {
+    return core;
+  }
+
+  @Override
+  @SuppressWarnings("squid:S2093" /* closing of real client handled by finalizeStateChange() */)
+  public void close() throws IOException {
+    try {
+      if (state == State.CLOSED) {
```

Comment:
How can you say? `close()` is part of the public `SolrClient` api. Anybody could be calling it at their own will. I have to protect against it.

Reply:
I didn't mean that `close()` couldn't be called multiple times. Just that the optimization (checking before we acquire the `lock`) is unnecessary imho since I can't think of a scenario where a client would call it multiple times on purpose, or at least enough times to have any kind of performance impact.

Reply:
Agreed. It just keep the code built the same way no matter the method.